### PR TITLE
G5K provider: support multiple queues in resa

### PR DESCRIPTION
**Pull request by asimonet**
This PR allows to use nodes from different queues in a deployment.

### CODE REVIEW

File: enos/provider/g5k.py [hunk 1]
```diff
@@ -218,56 +269,96 @@ def _check_nodes(self,
...
         if 'user' in job_info:
             self.user = job_info['user']
 
         # vlans information
-        job_sites = EX5.get_oargrid_job_oar_jobs(self.gridjob)
+        job_sites = []
+        for job_id in self.gridjobs:
+            job_sites.append(EX5.get_oargrid_job_oar_jobs(job_id))
+            
+        job_sites = [item for sublist in job_sites for item in sublist]
```
Comment (rcherrueau) [inline enos/provider/g5k.py]: A way to write job_sites in one go:
```python
job_sites = reduce(operator.add, map(EX5.get_oargrid_job_oar_jobs, self.gridjobs))

File: enos/provider/g5k.py [hunk 2]
```diff
@@ -218,56 +269,96 @@ def _check_nodes(self,
...
+        
         self.jobs = []
         self.vlans = []
         for (job_id, site) in job_sites:
             self.jobs.append((site, job_id))
-            vlan_id = EX5.get_oar_job_kavlan(job_id, site)
-            if vlan_id is not None:
-                self.vlans.append((site,
-                                   EX5.get_oar_job_kavlan(job_id, site)))
+
```
Comment (rcherrueau) [inline enos/provider/g5k.py]: I don't get why you rewrite the following snippet. I see nothing specific to multiple queues. Can you enlighten me on that?

File: enos/provider/g5k.py [hunk 3]
```diff
@@ -83,10 +84,10 @@ def init(self, config, calldir, force_deploy=False):
 
     def destroy(self, calldir, env):
         self._load_config(env['config'])
-        self.gridjob, _ = EX5.planning.get_job_by_name(self.config['name'])
-        if self.gridjob is not None:
-            EX5.oargriddel([self.gridjob])
-            logging.info("Killing the job %s" % self.gridjob)
+        self._get_job_ids()
+        
+        EX5.oargriddel(self.gridjobs)
```
Comment (asimonet) [inline enos/provider/g5k.py]: Good point, I'll fix it.

File: enos/provider/g5k.py [hunk 4]
```diff
@@ -165,37 +166,87 @@ def _make_reservation(self):
         criteria = {}
         # Actual criteria are :
         # - Number of node per site
+        queues = self._get_queues()
```
Comment (asimonet) [inline enos/provider/g5k.py]: But what about complexity??
In the case where we have two queues, that's four iterations instead of two!! That is twice as many iterations!!

File: enos/provider/g5k.py [hunk 5]
```diff
@@ -165,37 +166,87 @@ def _make_reservation(self):
...
+        for queue, clusters in queues.iteritems():
```
Comment (asimonet) [inline enos/provider/g5k.py]: This won't work, as I really need `clusters`.

File: enos/provider/g5k.py [hunk 6]
```diff
@@ -165,37 +166,87 @@ def _make_reservation(self):
...
+                self.gridjobs.append(gridjob)
+                logging.info("Using new oargrid job %s" % gridjob)
+            else:
+                logging.error("No oar job was created.")
+                sys.exit(26)
+
+
+    def _get_queues(self):
+        try:
+            return self.queues
```
Comment (asimonet) [inline enos/provider/g5k.py]: I propose to rename `self.queues` to `self._queues`.

File: enos/provider/g5k.py (hunk 4 above)
Comment (rcherrueau) [inline enos/provider/g5k.py]: :trollface:

File: enos/provider/g5k.py (hunk 5 above)
Comment (rcherrueau) [inline enos/provider/g5k.py]: OK got it.

File: enos/provider/g5k.py (hunk 6 above)
Comment (rcherrueau) [inline enos/provider/g5k.py]: 👍

**Comment by msimonin** [comment]
I would be careful before investing too much time into this.
I'm still not sure how frequent we would need this and what problem it solves. It looked like a workaround at a specific point of time where ressources were missing on g5k and new were available on another queue. 
Additionally I guess it requires some documentation update.

**Comment by asimonet** [comment]
msimonin:
− No investment is required, the code is already written 🙂
− It can be useful to try new clusters (3 new ones are coming to Lille, in addition to nova in Lyon)
− It allows to use the production queue
− I need it. I spent a weekend implementing this, and I don't want to lose it. I don't want to maintain it on a separate branch, because if the g5k provider changes in a few month, I will have to work on it for another weekend.
− It doesn't break anything, it doesn't alter any API, it doesn't require any particular configuration.
− I can add some documentation, but because of the above remark, I don't think it is necessary.

File: enos/provider/g5k.py [hunk 7]
```diff
@@ -165,35 +167,88 @@ def _make_reservation(self):
...
+            for queue, qc in criteria.iteritems():
+                if site in qc.keys():
+                    criteria[queue][site].append(vlan)
+                    break
+
+        # Make one reservation per queue
+        self.gridjobs = []
+        for queue in queues:
+            # Compute the specification for the reservation
+            name = "%s-%d" % (self.config['name'], len(self.gridjobs))
```
Comment (msimonin) [inline enos/provider/g5k.py]: The way the name is built should be factorized.

File: enos/provider/g5k.py [hunk 8]
```diff
@@ -165,35 +167,88 @@ def _make_reservation(self):
...
+                logging.info("Using new oargrid job %s" % gridjob)
+            else:
+                logging.error("No oar job was created.")
+                if len(self.gridjobs) > 0:
+                    EX5.oargriddel(self.gridjobs)
+                    logging.error("Deleting already submitted jobs: %s" % self.gridjobs)
+                sys.exit(26)
+
+
+    def _get_queues(self):
```
Comment (msimonin) [inline enos/provider/g5k.py]: It's a general idea for the providers.
There's an implicit here : `self.config` should be filled prior to the call. Personnally I don't like it and I would prefer passing the config as a argument. In addition this makes the method easier to test in isolation.

File: enos/provider/g5k.py [hunk 9]
```diff
@@ -165,35 +167,88 @@ def _make_reservation(self):
...
+            resources = EX5.get_resource_attributes('/sites/%s/clusters/%s' % (site, cluster))
+            queue = None
+            for q in QUEUE_PRIORITIES:
+                if q in resources['queues']:
+                    queue = q
+                    break
+
+            # This allows to use the 'admin' queue in last resort
+            if queue is None:
+                queue = cluster_queues[0]
```
Comment (msimonin) [inline enos/provider/g5k.py]: `cluster_queues` seems undefined ? I maybe miss something here.

File: enos/provider/g5k.py [hunk 10]
```diff
@@ -218,56 +273,88 @@ def _check_nodes(self,
...
-            raise Exception("Not enough nodes to continue")
+            raise Exception("Not enough nodes to continue (%d missing)"
+                    % (wanted_nodes - len(nodes)))
 
         return True
 
-    def _get_job(self):
-        """Get the hosts from an existing job (if any) or from a new job.
-        This will perform a reservation if necessary."""
+    def _get_job_ids(self):
```
Comment (msimonin) [inline enos/provider/g5k.py]: The same as `_get_queues`, I would pass the `config` and try to avoid any side effects (`self.gridjobs`). This is a design decision that has to be discussed and that goes beyond the current PR. More generally, many methods in the provider are only utils functions that could be extracted elsewhere. 
(cc @rcherrueau)

File: enos/provider/g5k.py [hunk 11]
```diff
@@ -218,56 +273,88 @@ def _check_nodes(self,
...
+        of job ids to `self.gridjobs` and returns it. If no job is found,
+        this method returns `None`. If the number of jobs found is more
+        than zero but less than the required number, this method quits
+        the program and prints the ids of the found jobs.
+        """
+        queues = self._get_queues()
+        jobs = []
+
+        for i in range(len(queues)):
+            name = "%s-%d" % (self.config['name'], i)
```
Comment (msimonin) [inline enos/provider/g5k.py]: see above (extract this)

File: enos/provider/g5k.py (hunk 10 above)
Comment (rcherrueau) [inline enos/provider/g5k.py]: +1 for purely functional util functions. Referential transparency helps reasoning and will help modularization if we decide to move such util functions in the util package.

File: enos/provider/g5k.py (hunk 7 above)
Comment (asimonet) [inline enos/provider/g5k.py]: 👍

File: enos/provider/g5k.py (hunk 9 above)
Comment (asimonet) [inline enos/provider/g5k.py]: You are not mistaken, I will fix this.

File: enos/provider/g5k.py (hunk 8 above)
Comment (asimonet) [inline enos/provider/g5k.py]: I agree but I think this should be a different PR, because either I change a the signature of a few methods (and then the class will be inconsistent) or I do it for all the methods (but then I touch many lines that are not related to this PR).

File: enos/provider/g5k.py (hunk 10 above)
Comment (asimonet) [inline enos/provider/g5k.py]: Same comment as before, I would rather keep this to a different PR.

File: enos/provider/g5k.py (hunk 11 above)
Comment (asimonet) [inline enos/provider/g5k.py]: 👍

**Comment by msimonin** [comment]
Up ? 

Beware there's something very strange about the above commits (@asimonet is co-authoring many commits :) )

**Comment by msimonin** [comment]
I will integrate the support after #112 is merged and make the feature testable.
